File: frontend/web/src/services/academicYearService.ts
import api from "../utils/client";

/* ================= TYPES ================= */

export type AcademicYear = {
  id: string;
  description?: string;
  batchYear?: string;
  startDate?: string;
  endDate?: string;
  insertBy?: string;
  updateBy?: string;
  insertOn?: string;
  updateOn?: string;
  status?: boolean;
};

export type CreateAcademicYearDto = {
  description?: string;
  batchYear?: string;
  startDate?: string;
  endDate?: string;
};

export type UpdateAcademicYearDto = {
  id: string;
  description?: string;
  batchYear?: string;
  startDate?: string;
  endDate?: string;
};

/* ================= CREATE ================= */

export const createAcademicYear = async (
  data: CreateAcademicYearDto
) => {
  const response = await api.post<AcademicYear>(
    "/AcademicYear",
    data
  );
  return response.data;
};

/* ================= GET ALL ================= */

export const getAcademicYears = async () => {
  const response = await api.get<AcademicYear[]>(
    "/AcademicYear"
  );
  return response.data;
};

/* ================= GET BY ID ================= */

export const getAcademicYearById = async (id: string) => {
  const response = await api.get<AcademicYear>(
    `/AcademicYear/${id}`
  );
  return response.data;
};

/* ================= UPDATE ================= */

export const updateAcademicYear = async (
  data: UpdateAcademicYearDto
) => {
  const response = await api.put<{ success: boolean }>(
    "/AcademicYear",
    data
  );
  return response.data;
};

/* ================= DELETE ================= */

export const deleteAcademicYear = async (id: string) => {
  const response = await api.delete<{ success: boolean }>(
    `/AcademicYear/${id}`
  );
  return response.data;
};

File: backend/UniversitySystem.Application/Services/AcademicYearService.cs
using KVFSU.Application.Interfaces.Common;

public class AcademicYearService : IAcademicYearService
{
    private readonly IAcademicYearRepository _repo;
    private readonly ICurrentUserService _currentUser;
    private readonly IAuditLogService _auditLog;

    public AcademicYearService(
        IAcademicYearRepository repo,
        ICurrentUserService currentUser,
        IAuditLogService auditLog)
    {
        _repo = repo;
        _currentUser = currentUser;
        _auditLog = auditLog;
    }

    // ✅ CREATE
    public async Task<AcademicYear> CreateAsync(CreateAcademicYearDto dto)
    {
        var entity = new AcademicYear
        {
            Id = Guid.NewGuid(),
            Description = dto.Description,
            StartDate = dto.StartDate,
            EndDate = dto.EndDate,
            InsertBy = _currentUser.Username,
            InsertOn = DateTime.UtcNow,
            Status = true,
            BatchYear = dto.BatchYear
        };

        var result = await _repo.CreateAsync(entity);

        await _auditLog.LogAsync("academic_years", result.Id, "INSERT", null, new
        {
            result.Description,
            result.BatchYear,
            result.StartDate,
            result.EndDate,
            result.InsertBy
        });

        return result;
    }

    // ✅ GET ALL
    public async Task<List<AcademicYear>> GetAllAsync()
    {
        return await _repo.GetAllAsync();
    }

    // ✅ GET BY ID
    public async Task<AcademicYear?> GetByIdAsync(Guid id)
    {
        return await _repo.GetByIdAsync(id);
    }

    // ✅ UPDATE
    public async Task<bool> UpdateAsync(UpdateAcademicYearDto dto)
    {
        var entity = await _repo.GetByIdAsync(dto.Id);

        if (entity == null)
            return false;

        var old = new { entity.Description, entity.StartDate, entity.EndDate };

        entity.Description = dto.Description;
        entity.BatchYear = dto.BatchYear;
        entity.StartDate = dto.StartDate;
        entity.EndDate = dto.EndDate;
        entity.UpdateBy = _currentUser.Username;
        entity.UpdateOn = DateTime.UtcNow;

        await _repo.UpdateAsync(entity);

        await _auditLog.LogAsync("academic_years", entity.Id, "UPDATE", old, new
        {
            entity.Description,
            entity.BatchYear,
            entity.StartDate,
            entity.EndDate,
            entity.UpdateBy
        });

        return true;
    }

    // ✅ SOFT DELETE
    public async Task<bool> DeleteAsync(Guid id)
    {
        var entity = await _repo.GetByIdAsync(id);

        if (entity == null)
            return false;

        var old = new { entity.Description, entity.Status, entity.BatchYear };

        entity.Status = false;
        entity.UpdateBy = _currentUser.Username;
        entity.UpdateOn = DateTime.UtcNow;

        await _repo.UpdateAsync(entity);

        await _auditLog.LogAsync("academic_years", entity.Id, "DELETE", old, null);

        return true;
    }
}
